File: chess_pipeline/pattern_detectors.py
"""Chess tactical-pattern detectors.

Pure functions over python-chess boards. Each detector answers:
"does `move`, played from `board`, create the named pattern soundly?"

Sound = the moved piece cannot be immediately captured at a profit
(checked against *legal* replies, so pins and check obligations are
respected automatically).

Reusable across projects: no game-format or app-specific imports.
"""

from __future__ import annotations

import chess

PIECE_VALUES = {
    chess.PAWN: 1,
    chess.KNIGHT: 3,
    chess.BISHOP: 3,
    chess.ROOK: 5,
    chess.QUEEN: 9,
    chess.KING: 100,
}

SLIDERS = (chess.BISHOP, chess.ROOK, chess.QUEEN)


def piece_value(piece: chess.Piece | None) -> int:
    return PIECE_VALUES[piece.piece_type] if piece else 0


def is_move_safe(board_after: chess.Board, square: chess.Square) -> bool:
    """True if the piece on `square` cannot be captured at a profit by any
    legal reply. Equal-value trades by an undefended capture also count
    as unsafe (the tactic would just be a swap, not a pattern win)."""
    mover = board_after.piece_at(square)
    if mover is None:
        return True
    mover_val = piece_value(mover)
    defenders = board_after.attackers(mover.color, square)

    for reply in board_after.legal_moves:
        if reply.to_square != square or not board_after.is_capture(reply):
            continue
        capturer = board_after.piece_at(reply.from_square)
        capturer_val = piece_value(capturer)
        if capturer_val < mover_val:
            return False  # cheaper piece takes: always a loss
        if not defenders:
            return False  # undefended: any capture loses the piece
    return True


def _ray_squares(from_sq: chess.Square, direction: tuple[int, int]):
    """Yield squares stepping (df, dr) from from_sq until the board edge."""
    f, r = chess.square_file(from_sq), chess.square_rank(from_sq)
    df, dr = direction
    while True:
        f, r = f + df, r + dr
        if not (0 <= f <= 7 and 0 <= r <= 7):
            return
        yield chess.square(f, r)


def _slider_directions(piece_type: chess.PieceType) -> list[tuple[int, int]]:
    diag = [(1, 1), (1, -1), (-1, 1), (-1, -1)]
    orth = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    if piece_type == chess.BISHOP:
        return diag
    if piece_type == chess.ROOK:
        return orth
    if piece_type == chess.QUEEN:
        return diag + orth
    return []


def _line_pairs(board: chess.Board, slider_sq: chess.Square):
    """For each ray of the slider on slider_sq, yield (front, behind):
    the first two occupied squares along the ray."""
    slider = board.piece_at(slider_sq)
    if slider is None:
        return
    for direction in _slider_directions(slider.piece_type):
        front = None
        for sq in _ray_squares(slider_sq, direction):
            if board.piece_at(sq) is None:
                continue
            if front is None:
                front = sq
            else:
                yield front, sq
                break


def detect_fork(board: chess.Board, move: chess.Move) -> bool:
    """Moved piece attacks >=2 enemy targets at once, each of which is
    the king (check), worth more than the mover, or undefended."""
    if move not in board.legal_moves:
        return False
    after = board.copy()
    after.push(move)
    sq = move.to_square
    mover = after.piece_at(sq)
    if mover is None or not is_move_safe(after, sq):
        return False

    mover_val = piece_value(mover)
    enemy = not mover.color
    targets = 0
    for t in after.attacks(sq):
        victim = after.piece_at(t)
        if victim is None or victim.color != enemy:
            continue
        if victim.piece_type == chess.KING:
            targets += 1
        elif piece_value(victim) > mover_val:
            targets += 1
        elif piece_value(victim) >= 3 and not after.attackers(enemy, t):
            targets += 1
    return targets >= 2


def detect_pin(board: chess.Board, move: chess.Move) -> bool:
    """Move places a slider so an enemy piece is stuck in front of its
    king (absolute pin) or in front of a heavier piece (Q/R behind)."""
    if move not in board.legal_moves:
        return False
    after = board.copy()
    after.push(move)
    sq = move.to_square
    mover = after.piece_at(sq)
    if mover is None or mover.piece_type not in SLIDERS:
        return False
    if not is_move_safe(after, sq):
        return False

    enemy = not mover.color
    for front, behind in _line_pairs(after, sq):
        p1, p2 = after.piece_at(front), after.piece_at(behind)
        if p1.color != enemy or p2.color != enemy:
            continue
        if p1.piece_type == chess.KING:
            continue  # that ordering is a skewer, not a pin
        if piece_value(p1) < 3:
            continue  # pinned pawns are not worth teaching as a pin
        # Fake pin: the "pinned" piece can simply capture the pinner
        # (e.g. bishop "pinning" a bishop on the same diagonal).
        if sq in after.attacks(front):
            pinner_defended = bool(after.attackers(mover.color, sq))
            if not pinner_defended or piece_value(p1) <= piece_value(mover):
                continue
        absolute = p2.piece_type == chess.KING
        relative = piece_value(p2) >= 5 and piece_value(p2) > piece_value(p1)
        if absolute or relative:
            # The pin must bite: pinned piece is attacked by the slider
            # and cannot profitably take it (slider already checked safe),
            # and the pinned piece is worth winning or paralysed.
            if absolute and after.is_pinned(enemy, front):
                return True
            if relative:
                return True
    return False


def detect_skewer(board: chess.Board, move: chess.Move) -> bool:
    """Move attacks a high-value enemy piece (K or Q) in line with a
    lesser piece behind it: when the front piece moves, the back one falls."""
    if move not in board.legal_moves:
        return False
    after = board.copy()
    after.push(move)
    sq = move.to_square
    mover = after.piece_at(sq)
    if mover is None or mover.piece_type not in SLIDERS:
        return False
    if not is_move_safe(after, sq):
        return False

    enemy = not mover.color
    mover_val = piece_value(mover)
    for front, behind in _line_pairs(after, sq):
        p1, p2 = after.piece_at(front), after.piece_at(behind)
        if p1.color != enemy or p2.color != enemy:
            continue
        front_forced = p1.piece_type == chess.KING or (
            p1.piece_type == chess.QUEEN and mover_val < piece_value(p1)
        )
        if not front_forced:
            continue
        if not (3 <= piece_value(p2) < piece_value(p1)):
            continue
        # The loot must actually be winnable once the front piece flees:
        # undefended (not counting the fleeing piece), or worth more
        # than the attacker.
        defenders = after.attackers(enemy, behind) & ~chess.BB_SQUARES[front]
        if not defenders or piece_value(p2) > mover_val:
            return True
    return False


def detect_discovered_attack(board: chess.Board, move: chess.Move) -> bool:
    """Moving one piece uncovers a friendly slider's attack on the enemy
    king (discovered check) or a heavy piece, while the moved piece itself
    also does something (captures, checks, or attacks material)."""
    if move not in board.legal_moves:
        return False
    color = board.turn
    enemy = not color
    after = board.copy()
    after.push(move)

    if not is_move_safe(after, move.to_square):
        return False

    best_discovered = 0  # 100 = king, 9 = queen, 5 = rook
    for slider_sq in (
        list(after.pieces(chess.BISHOP, color))
        + list(after.pieces(chess.ROOK, color))
        + list(after.pieces(chess.QUEEN, color))
    ):
        if slider_sq == move.to_square:
            continue  # the mover itself is not the discovered piece
        # The vacated square must sit on the line between slider and target.
        for target in after.attacks(slider_sq):
            victim = after.piece_at(target)
            if victim is None or victim.color != enemy:
                continue
            if victim.piece_type != chess.KING and piece_value(victim) < 5:
                continue
            if move.from_square not in chess.SquareSet.between(slider_sq, target):
                continue
            if target in board.attacks(slider_sq):
                continue  # was already attacked: nothing was discovered
            # Refutable: the attacked piece just captures the slider.
            if slider_sq in after.attacks(target) and not after.attackers(
                color, slider_sq
            ):
                continue
            best_discovered = max(best_discovered, piece_value(victim))
        if best_discovered >= 9:
            break
    if not best_discovered:
        return False

    # A revealed attack on king or queen forces a response by itself.
    if best_discovered >= 9:
        return True

    # A revealed attack on a rook needs the mover to add its own threat
    # (a true double attack), or the rook simply steps away.
    if board.is_capture(move):
        return True
    if after.is_check():
        return True
    mover_val = piece_value(after.piece_at(move.to_square))
    for t in after.attacks(move.to_square):
        victim = after.piece_at(t)
        if victim and victim.color == enemy and piece_value(victim) > mover_val:
            return True
        if victim and victim.color == enemy and piece_value(victim) >= 3 \
                and not after.attackers(enemy, t):
            return True
    return False


def detect_double_check(board: chess.Board, move: chess.Move) -> bool:
    """Two pieces give check simultaneously."""
    if move not in board.legal_moves:
        return False
    after = board.copy()
    after.push(move)
    return len(after.checkers()) == 2


def detect_back_rank_mate(board: chess.Board, move: chess.Move) -> bool:
    """Checkmate delivered by a heavy piece along the enemy back rank,
    the king walled in on that rank."""
    if move not in board.legal_moves:
        return False
    after = board.copy()
    after.push(move)
    if not after.is_checkmate():
        return False

    enemy = after.turn  # side that just got mated
    king_sq = after.king(enemy)
    back_rank = 7 if enemy == chess.BLACK else 0
    if chess.square_rank(king_sq) != back_rank:
        return False

    for checker_sq in after.checkers():
        checker = after.piece_at(checker_sq)
        if checker.piece_type in (chess.ROOK, chess.QUEEN) and (
            chess.square_rank(checker_sq) == back_rank
        ):
            return True
    return False


DETECTORS = {
    "fork": detect_fork,
    "pin": detect_pin,
    "skewer": detect_skewer,
    "discovered_attack": detect_discovered_attack,
    "double_check": detect_double_check,
    "back_rank_mate": detect_back_rank_mate,
}

File: lib/patternDetectors.ts
/**
 * Chess tactical-pattern detectors (TypeScript port of
 * chess_pipeline/pattern_detectors.py — keep the two in sync).
 *
 * Pure functions over FEN strings. Each detector answers:
 * "does `moveUci`, played from `fen`, create the named pattern soundly?"
 *
 * Sound = the moved piece cannot be immediately captured at a profit
 * (checked against *legal* replies, so pins and check obligations are
 * respected automatically).
 *
 * Reusable across projects: no game-format or app-specific imports.
 */

import { Chess, Color, PieceSymbol, Square } from 'chess.js'

export const PIECE_VALUES: Record<PieceSymbol, number> = {
  p: 1, n: 3, b: 3, r: 5, q: 9, k: 100,
}

const SLIDERS: PieceSymbol[] = ['b', 'r', 'q']

const DIAG: Array<[number, number]> = [[1, 1], [1, -1], [-1, 1], [-1, -1]]
const ORTH: Array<[number, number]> = [[1, 0], [-1, 0], [0, 1], [0, -1]]

// ── Square helpers ────────────────────────────────────────────────
const FILES = 'abcdefgh'

function fileOf(s: Square): number { return FILES.indexOf(s[0]) }
function rankOf(s: Square): number { return Number(s[1]) - 1 }

function makeSquare(file: number, rank: number): Square | null {
  if (file < 0 || file > 7 || rank < 0 || rank > 7) return null
  return (FILES[file] + String(rank + 1)) as Square
}

function* raySquares(from: Square, [df, dr]: [number, number]): Generator<Square> {
  let f = fileOf(from) + df
  let r = rankOf(from) + dr
  while (true) {
    const s = makeSquare(f, r)
    if (s === null) return
    yield s
    f += df
    r += dr
  }
}

function sliderDirections(type: PieceSymbol): Array<[number, number]> {
  if (type === 'b') return DIAG
  if (type === 'r') return ORTH
  if (type === 'q') return [...DIAG, ...ORTH]
  return []
}

/** Squares strictly between two aligned squares; empty array if unaligned. */
export function squaresBetween(a: Square, b: Square): Square[] {
  const df = fileOf(b) - fileOf(a)
  const dr = rankOf(b) - rankOf(a)
  if (df === 0 && dr === 0) return []
  if (df !== 0 && dr !== 0 && Math.abs(df) !== Math.abs(dr)) return []
  const step: [number, number] = [Math.sign(df), Math.sign(dr)]
  const out: Square[] = []
  for (const s of raySquares(a, step)) {
    if (s === b) break
    out.push(s)
  }
  return out
}

// ── Board helpers ─────────────────────────────────────────────────
function other(color: Color): Color { return color === 'w' ? 'b' : 'w' }

function pieceValue(piece: { type: PieceSymbol } | undefined | null): number {
  return piece ? PIECE_VALUES[piece.type] : 0
}

function* allPieces(game: Chess): Generator<[Square, { type: PieceSymbol; color: Color }]> {
  for (let r = 0; r < 8; r++) {
    for (let f = 0; f < 8; f++) {
      const s = makeSquare(f, r)!
      const p = game.get(s)
      if (p) yield [s, p]
    }
  }
}

function findKing(game: Chess, color: Color): Square | null {
  for (const [s, p] of allPieces(game)) {
    if (p.type === 'k' && p.color === color) return s
  }
  return null
}

/** True if the piece on `square` attacks `target` (pseudo-legal attack). */
function attacksSquare(game: Chess, square: Square, target: Square): boolean {
  const piece = game.get(square)
  if (!piece) return false
  return game.attackers(target, piece.color).includes(square)
}

/**
 * True if the piece on `square` cannot be captured at a profit by any
 * legal reply. Equal-value trades by an undefended capture also count
 * as unsafe (the tactic would just be a swap, not a pattern win).
 */
export function isMoveSafe(gameAfter: Chess, square: Square): boolean {
  const mover = gameAfter.get(square)
  if (!mover) return true
  const moverVal = PIECE_VALUES[mover.type]
  const defenders = gameAfter.attackers(square, mover.color)

  for (const reply of gameAfter.moves({ verbose: true })) {
    if (reply.to !== square || !reply.captured) continue
    const capturerVal = PIECE_VALUES[reply.piece]
    if (capturerVal < moverVal) return false // cheaper piece takes: always a loss
    if (defenders.length === 0) return false // undefended: any capture loses
  }
  return true
}

/** For each ray of the slider on sliderSq, yield [front, behind]:
 *  the first two occupied squares along the ray. */
function* linePairs(game: Chess, sliderSq: Square): Generator<[Square, Square]> {
  const slider = game.get(sliderSq)
  if (!slider) return
  for (const dir of sliderDirections(slider.type)) {
    let front: Square | null = null
    for (const s of raySquares(sliderSq, dir)) {
      if (!game.get(s)) continue
      if (front === null) {
        front = s
      } else {
        yield [front, s]
        break
      }
    }
  }
}

/** Absolute pin: the piece on `square` shields its own king from an
 *  enemy slider along a straight or diagonal line. */
function isAbsolutelyPinned(game: Chess, square: Square): boolean {
  const piece = game.get(square)
  if (!piece) return false
  const kingSq = findKing(game, piece.color)
  if (!kingSq || kingSq === square) return false

  const df = fileOf(square) - fileOf(kingSq)
  const dr = rankOf(square) - rankOf(kingSq)
  if (df !== 0 && dr !== 0 && Math.abs(df) !== Math.abs(dr)) return false

  // king -> square must be unobstructed
  for (const s of squaresBetween(kingSq, square)) {
    if (game.get(s)) return false
  }
  // continue past the piece: first occupied square must be an enemy
  // slider that moves along this direction
  const step: [number, number] = [Math.sign(df), Math.sign(dr)]
  const diagonal = step[0] !== 0 && step[1] !== 0
  for (const s of raySquares(square, step)) {
    const p = game.get(s)
    if (!p) continue
    if (p.color === piece.color) return false
    if (p.type === 'q') return true
    return diagonal ? p.type === 'b' : p.type === 'r'
  }
  return false
}

// ── Move application ──────────────────────────────────────────────
interface MoveContext {
  before: Chess
  after: Chess
  from: Square
  to: Square
  color: Color           // side that moved
  wasCapture: boolean
}

/** Apply a UCI move if legal; null otherwise. */
function applyMove(fen: string, moveUci: string): MoveContext | null {
  let before: Chess
  try {
    before = new Chess(fen)
  } catch {
    return null
  }
  const from = moveUci.slice(0, 2) as Square
  const to = moveUci.slice(2, 4) as Square
  const promotion = moveUci.slice(4, 5) || undefined

  const legal = before.moves({ verbose: true }).find(
    m => m.from === from && m.to === to && (m.promotion ?? '') === (promotion ?? '')
  )
  if (!legal) return null

  const after = new Chess(fen)
  after.move({ from, to, promotion })
  return { before, after, from, to, color: before.turn(), wasCapture: Boolean(legal.captured) }
}

// ── Detectors ─────────────────────────────────────────────────────
export type PatternDetector = (fen: string, moveUci: string) => boolean

/** Moved piece attacks >=2 enemy targets at once, each of which is
 *  the king (check), worth more than the mover, or undefended. */
export function detectFork(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { after, to } = ctx
  const mover = after.get(to)
  if (!mover || !isMoveSafe(after, to)) return false

  const moverVal = PIECE_VALUES[mover.type]
  const enemy = other(mover.color)
  let targets = 0
  for (const [t, victim] of allPieces(after)) {
    if (victim.color !== enemy) continue
    if (!attacksSquare(after, to, t)) continue
    if (victim.type === 'k') {
      targets += 1
    } else if (PIECE_VALUES[victim.type] > moverVal) {
      targets += 1
    } else if (PIECE_VALUES[victim.type] >= 3 && after.attackers(t, enemy).length === 0) {
      targets += 1
    }
  }
  return targets >= 2
}

/** Move places a slider so an enemy piece is stuck in front of its
 *  king (absolute pin) or in front of a heavier piece (Q/R behind). */
export function detectPin(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { after, to } = ctx
  const mover = after.get(to)
  if (!mover || !SLIDERS.includes(mover.type)) return false
  if (!isMoveSafe(after, to)) return false

  const enemy = other(mover.color)
  for (const [front, behind] of linePairs(after, to)) {
    const p1 = after.get(front)!
    const p2 = after.get(behind)!
    if (p1.color !== enemy || p2.color !== enemy) continue
    if (p1.type === 'k') continue // that ordering is a skewer, not a pin
    if (PIECE_VALUES[p1.type] < 3) continue // pinned pawns are not worth teaching
    // Fake pin: the "pinned" piece can simply capture the pinner
    // (e.g. bishop "pinning" a bishop on the same diagonal).
    if (attacksSquare(after, front, to)) {
      const pinnerDefended = after.attackers(to, mover.color).length > 0
      if (!pinnerDefended || PIECE_VALUES[p1.type] <= PIECE_VALUES[mover.type]) continue
    }
    const absolute = p2.type === 'k'
    const relative = PIECE_VALUES[p2.type] >= 5 && PIECE_VALUES[p2.type] > PIECE_VALUES[p1.type]
    if (absolute && isAbsolutelyPinned(after, front)) return true
    if (relative) return true
  }
  return false
}

/** Move attacks a high-value enemy piece (K or Q) in line with a
 *  lesser piece behind it: when the front piece moves, the back one falls. */
export function detectSkewer(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { after, to } = ctx
  const mover = after.get(to)
  if (!mover || !SLIDERS.includes(mover.type)) return false
  if (!isMoveSafe(after, to)) return false

  const enemy = other(mover.color)
  const moverVal = PIECE_VALUES[mover.type]
  for (const [front, behind] of linePairs(after, to)) {
    const p1 = after.get(front)!
    const p2 = after.get(behind)!
    if (p1.color !== enemy || p2.color !== enemy) continue
    const frontForced =
      p1.type === 'k' || (p1.type === 'q' && moverVal < PIECE_VALUES[p1.type])
    if (!frontForced) continue
    if (!(PIECE_VALUES[p2.type] >= 3 && PIECE_VALUES[p2.type] < PIECE_VALUES[p1.type])) continue
    // The loot must actually be winnable once the front piece flees:
    // undefended (not counting the fleeing piece), or worth more
    // than the attacker.
    const defenders = after.attackers(behind, enemy).filter(s => s !== front)
    if (defenders.length === 0 || PIECE_VALUES[p2.type] > moverVal) return true
  }
  return false
}

/** Moving one piece uncovers a friendly slider's attack on the enemy
 *  king (discovered check) or a heavy piece, while the moved piece itself
 *  also does something (captures, checks, or attacks material). */
export function detectDiscoveredAttack(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { before, after, from, to, color, wasCapture } = ctx
  const enemy = other(color)

  if (!isMoveSafe(after, to)) return false

  let bestDiscovered = 0 // 100 = king, 9 = queen, 5 = rook
  outer:
  for (const [sliderSq, slider] of allPieces(after)) {
    if (slider.color !== color || !SLIDERS.includes(slider.type)) continue
    if (sliderSq === to) continue // the mover itself is not the discovered piece
    for (const [target, victim] of allPieces(after)) {
      if (victim.color !== enemy) continue
      if (victim.type !== 'k' && PIECE_VALUES[victim.type] < 5) continue
      if (!attacksSquare(after, sliderSq, target)) continue
      // The vacated square must sit on the line between slider and target.
      if (!squaresBetween(sliderSq, target).includes(from)) continue
      if (attacksSquare(before, sliderSq, target)) continue // already attacked
      // Refutable: the attacked piece just captures the slider.
      if (
        attacksSquare(after, target, sliderSq) &&
        after.attackers(sliderSq, color).length === 0
      ) continue
      bestDiscovered = Math.max(bestDiscovered, PIECE_VALUES[victim.type])
      if (bestDiscovered >= 9) break outer
    }
  }
  if (!bestDiscovered) return false

  // A revealed attack on king or queen forces a response by itself.
  if (bestDiscovered >= 9) return true

  // A revealed attack on a rook needs the mover to add its own threat
  // (a true double attack), or the rook simply steps away.
  if (wasCapture) return true
  if (after.isCheck()) return true
  const moverVal = pieceValue(after.get(to))
  for (const [t, victim] of allPieces(after)) {
    if (victim.color !== enemy) continue
    if (!attacksSquare(after, to, t)) continue
    if (PIECE_VALUES[victim.type] > moverVal) return true
    if (PIECE_VALUES[victim.type] >= 3 && after.attackers(t, enemy).length === 0) return true
  }
  return false
}

/** Two pieces give check simultaneously. */
export function detectDoubleCheck(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { after, color } = ctx
  const kingSq = findKing(after, other(color))
  if (!kingSq) return false
  return after.attackers(kingSq, color).length === 2
}

/** Checkmate delivered by a heavy piece along the enemy back rank,
 *  the king walled in on that rank. */
export function detectBackRankMate(fen: string, moveUci: string): boolean {
  const ctx = applyMove(fen, moveUci)
  if (!ctx) return false
  const { after, color } = ctx
  if (!after.isCheckmate()) return false

  const enemy = other(color) // side that just got mated
  const kingSq = findKing(after, enemy)
  if (!kingSq) return false
  const backRank = enemy === 'b' ? 7 : 0
  if (rankOf(kingSq) !== backRank) return false

  for (const checkerSq of after.attackers(kingSq, color)) {
    const checker = after.get(checkerSq)!
    if ((checker.type === 'r' || checker.type === 'q') && rankOf(checkerSq) === backRank) {
      return true
    }
  }
  return false
}

export const DETECTORS: Record<string, PatternDetector> = {
  fork: detectFork,
  pin: detectPin,
  skewer: detectSkewer,
  discovered_attack: detectDiscoveredAttack,
  double_check: detectDoubleCheck,
  back_rank_mate: detectBackRankMate,
}

/** All detector-positive moves for `pattern` in the position. */
export function patternMoves(fen: string, pattern: string): string[] {
  const detector = DETECTORS[pattern]
  if (!detector) return []
  const game = new Chess(fen)
  const hits: string[] = []
  for (const m of game.moves({ verbose: true })) {
    const uci = `${m.from}${m.to}${m.promotion ?? ''}`
    if (detector(fen, uci)) hits.push(uci)
  }
  return hits
}
